File: app/src/main/java/cn/nicolite/huthelper/presenter/UserInfoPresenter.java
package cn.nicolite.huthelper.presenter;

import android.Manifest;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;

import com.yanzhenjie.permission.AndPermission;
import com.yanzhenjie.permission.PermissionListener;

import java.io.ByteArrayOutputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import cn.nicolite.huthelper.base.BasePresenter;
import cn.nicolite.huthelper.db.dao.UserDao;
import cn.nicolite.huthelper.model.bean.HttpResult;
import cn.nicolite.huthelper.model.bean.UploadImages;
import cn.nicolite.huthelper.model.bean.User;
import cn.nicolite.huthelper.network.APIUtils;
import cn.nicolite.huthelper.network.exception.ExceptionEngine;
import cn.nicolite.huthelper.utils.EncryptUtils;
import cn.nicolite.huthelper.view.activity.UserInfoActivity;
import cn.nicolite.huthelper.view.iview.IUserInfoView;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

/**
 * UserInfoPresenter
 * Created by nicolite on 17-10-28.
 */

public class UserInfoPresenter extends BasePresenter<IUserInfoView, UserInfoActivity> {
    public UserInfoPresenter(IUserInfoView view, UserInfoActivity activity) {
        super(view, activity);
    }

    public void showUserData() {

        User user = configure.getUser();

        if (user == null) {
            if (getView() != null) {
                getView().showMessage("获取当前登录用户失败，请重新登录！");
            }
            return;
        }
        getView().showUserInfo(user);
    }

    public void uploadAvatar(final Bitmap bitmap) {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, outputStream);
        byte[] bytes = outputStream.toByteArray();
        RequestBody requestBody = RequestBody.create(MediaType.parse("img/jpeg"), bytes);
        MultipartBody.Part file = MultipartBody.Part.createFormData("file", "01.jpg", requestBody);

        if (getView() != null) {
            getView().showMessage("头像上传中！");
        }

        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM", Locale.CHINA);
        String date = simpleDateFormat.format(new Date());
        String env = EncryptUtils.SHA1(configure.getStudentKH() + configure.getAppRememberCode() + date);

        APIUtils.INSTANCE
                .getUploadAPI()
                .uploadImages(configure.getStudentKH(), configure.getAppRememberCode(), env, 3, file)
                .compose(getActivity().<UploadImages>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<UploadImages>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        if (getView() != null) {
                            getView().showLoading();
                        }
                    }

                    @Override
                    public void onNext(UploadImages uploadImages) {
                        if (getView() != null) {
                            getView().closeLoading();
                            if (uploadImages.getCode() == 200) {
                                getView().showMessage("修改成功!");
                                User user = configure.getUser();
                                user.setHead_pic_thumb(uploadImages.getData());
                                user.setHead_pic(uploadImages.getData_original());
                                daoSession.getUserDao().update(user);
                                getView().changeAvatarSuccess(bitmap);
                            } else {
                                getView().showMessage("修改失败!");
                            }
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (getView() != null) {
                            getView().closeLoading();
                            getView().showMessage(ExceptionEngine.handleException(e).getMsg());
                        }
                    }

                    @Override
                    public void onComplete() {

                    }
                });

    }

    public void changAvatar() {
        AndPermission
                .with(getActivity())
                .requestCode(100)
                .permission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                .callback(new PermissionListener() {
                    @Override
                    public void onSucceed(int requestCode, @NonNull List<String> grantPermissions) {
                        if (getView() != null) {
                            if (requestCode == 100) {
                                getView().changeAvatar();
                            }
                        }
                    }

                    @Override
                    public void onFailed(int requestCode, @NonNull List<String> deniedPermissions) {
                        if (getView() != null) {
                            getView().showMessage("获取权限失败，请授予文件读写权限！");
                        }
                    }
                })
                .start();
    }

    public void changeUserName(final String userName) {

        if (getView() != null) {
            getView().showMessage("昵称修改中！");
        }

        APIUtils.INSTANCE
                .getUserAPI()
                .changeUsername(configure.getStudentKH(), configure.getAppRememberCode(), userName)
                .compose(getActivity().<HttpResult>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<HttpResult>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(HttpResult httpResult) {
                        if (getView() != null) {
                            if (httpResult.getCode() == 200) {
                                UserDao userDao = daoSession.getUserDao();
                                User user = configure.getUser();
                                user.setUsername(userName);
                                userDao.update(user);

                                getView().showMessage("修改成功!");
                            } else {
                                getView().showMessage("修改失败，code：" + httpResult.getCode()
                                        + "，" + httpResult.getMsg());
                            }
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (getView() != null) {
                            getView().showMessage(ExceptionEngine.handleException(e).getMsg());
                        }
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    //更换个人签名
    public void changeBio(final String bio) {

        APIUtils.INSTANCE
                .getUserAPI()
                .changeBio(configure.getStudentKH(), configure.getAppRememberCode(), bio)
                .compose(getActivity().<HttpResult>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<HttpResult>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(HttpResult httpResult) {
                        if (getView() != null) {
                            if (httpResult.getCode() == 200) {
                                UserDao userDao = daoSession.getUserDao();
                                User user = configure.getUser();
                                user.setBio(bio);
                                userDao.update(user);

                                getView().showMessage("修改成功!");
                            } else {
                                getView().showMessage("修改失败，" + httpResult.getCode());
                            }
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (getView() != null) {
                            getView().showMessage(ExceptionEngine.handleException(e).getMsg());
                        }
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/cn/nicolite/huthelper/presenter/UserInfoPresenter.kt
package cn.nicolite.huthelper.presenter

import android.Manifest
import android.graphics.Bitmap
import cn.nicolite.huthelper.kBase.BasePresenter
import cn.nicolite.huthelper.model.bean.HttpResult
import cn.nicolite.huthelper.model.bean.UploadImages
import cn.nicolite.huthelper.network.APIUtils
import cn.nicolite.huthelper.network.exception.ExceptionEngine
import cn.nicolite.huthelper.utils.EncryptUtils
import cn.nicolite.huthelper.view.fragment.UserInfoFragment
import cn.nicolite.huthelper.view.iview.IUserInfoView
import com.yanzhenjie.permission.AndPermission
import com.yanzhenjie.permission.PermissionListener
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.*

/**
 * UserInfoPresenter
 * Created by nicolite on 17-10-28.
 */

class UserInfoPresenter(iView: IUserInfoView, view: UserInfoFragment) : BasePresenter<IUserInfoView, UserInfoFragment>(iView, view) {

    fun showUserData() {
        val user = configure.user
        if (user == null) {
            getIView()?.showMessage("获取当前登录用户失败，请重新登录！")
        } else {
            getIView()?.showUserInfo(user)
        }
    }

    fun uploadAvatar(bitmap: Bitmap) {
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, outputStream)
        val bytes = outputStream.toByteArray()
        val requestBody = RequestBody.create(MediaType.parse("img/jpeg"), bytes)
        val file = MultipartBody.Part.createFormData("file", "01.jpg", requestBody)

        getIView()?.showMessage("头像上传中！")

        val simpleDateFormat = SimpleDateFormat("yyyy-MM", Locale.CHINA)
        val date = simpleDateFormat.format(Date())
        val env = EncryptUtils.SHA1(configure.studentKH + configure.appRememberCode + date)

        APIUtils
                .getUploadAPI()
                .uploadImages(configure.studentKH, configure.appRememberCode, env, 3, file)
                .compose(getView()?.bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<UploadImages> {
                    override fun onSubscribe(d: Disposable) {
                        getIView()?.showLoading()
                    }

                    override fun onNext(uploadImages: UploadImages) {
                        getIView()?.closeLoading()
                        if (uploadImages.code == 200) {
                            getIView()?.showMessage("修改成功!")
                            val user = configure.user
                            user.head_pic_thumb = uploadImages.data
                            user.head_pic = uploadImages.data_original
                            daoSession.userDao.update(user)
                            getIView()?.changeAvatarSuccess(bitmap)
                        } else {
                            getIView()?.showMessage("修改失败!")
                        }
                    }

                    override fun onError(e: Throwable) {
                        getIView()?.closeLoading()
                        getIView()?.showMessage(ExceptionEngine.handleException(e).msg)
                    }

                    override fun onComplete() {

                    }
                })

    }

    fun changAvatar() {
        AndPermission
                .with(fragment!!)
                .requestCode(100)
                .permission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                .callback(object : PermissionListener {
                    override fun onSucceed(requestCode: Int, grantPermissions: List<String>) {
                        if (requestCode == 100) {
                            getIView()?.changeAvatar()
                        }
                    }

                    override fun onFailed(requestCode: Int, deniedPermissions: List<String>) {
                        getIView()?.showMessage("获取权限失败，请授予文件读写权限！")
                    }
                })
                .start()
    }

    fun changeUserName(userName: String) {

        getIView()?.showMessage("昵称修改中！")

        APIUtils
                .getUserAPI()
                .changeUsername(configure.studentKH, configure.appRememberCode, userName)
                .compose(getView()?.bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<HttpResult<*>> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(httpResult: HttpResult<*>) {
                        if (httpResult.code == 200) {
                            val userDao = daoSession.userDao
                            val user = configure.user
                            user.username = userName
                            userDao.update(user)
                            getIView()?.showMessage("修改成功!")
                        } else {
                            getIView()?.showMessage("修改失败，code： ${httpResult.code}，${httpResult.msg}")
                        }
                    }

                    override fun onError(e: Throwable) {
                        getIView()?.showMessage(ExceptionEngine.handleException(e).msg)
                    }

                    override fun onComplete() {

                    }
                })
    }

    //更换个人签名
    fun changeBio(bio: String) {

        APIUtils
                .getUserAPI()
                .changeBio(configure.studentKH, configure.appRememberCode, bio)
                .compose(getView()?.bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<HttpResult<*>> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(httpResult: HttpResult<*>) {
                        if (httpResult.code == 200) {
                            val userDao = daoSession.userDao
                            val user = configure.user
                            user.bio = bio
                            userDao.update(user)
                            getIView()?.showMessage("修改成功!")
                        } else {
                            getIView()?.showMessage("修改失败，${httpResult.code}，${httpResult.msg}")
                        }
                    }

                    override fun onError(e: Throwable) {
                        getIView()?.showMessage(ExceptionEngine.handleException(e).msg)
                    }

                    override fun onComplete() {

                    }
                })
    }
}
